# GreedyGame Technical Assisment for Data Analyst Role

# GreedyGames Assignment
I have divided the Assigmnet into two parts. 
1. The First Part would directly Answer the question Asked, with the help od data manupulation done with the help of Python.
2. The Second part would Find some hidden insights from the data with the help of data visualization done on Power BI.

# Part 1st

## Some Basic Analysis

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
gg_df = pd.read_csv('gg_data.csv')
gg_df.head()

Date                     App                    Ad_Group       State  \
0  08-09-2020  (SDK X) Cricket Mazza   Cricket Mazza Banner (AdX)  Unresolved   
1  08-09-2020  (SDK X) Cricket Mazza                       320x50  Unresolved   
2  08-09-2020  (SDK X) Cricket Mazza               Dream11_320x50  Unresolved   
3  08-09-2020  (SDK X) Cricket Mazza                  My11_320x50  Unresolved   
4  06-09-2020  (SDK X) Cricket Mazza   Cricket Mazza Banner (AdX)  Unresolved   

   Revenue  Ad_Requests  Ad_Responses  Fill_Rate  Impressions  Render_Rate  \
0      0.0      3932330       3926253       99.9      3849749         98.1   
1      0.0      3932330           119       99.9          118         99.2   
2      0.0      3932330            96       99.9           93         96.9   
3      0.0      3932330           111       99.9          110         99.1   
4      0.0      3839002       3834328       99.9      3785127         98.7   

   1st_Clicks  2nd_Clicks  
0       10199           0  
1           2           2  
2           2           2  
3           2           2  
4        9492           0

In [3]:
gg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48293 entries, 0 to 48292
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Date          48293 non-null  object 
 1   App           48270 non-null  object 
 2   Ad_Group      44300 non-null  object 
 3   State         48293 non-null  object 
 4   Revenue       48293 non-null  float64
 5   Ad_Requests   48293 non-null  int64  
 6   Ad_Responses  48293 non-null  int64  
 7   Fill_Rate     48293 non-null  float64
 8   Impressions   48293 non-null  int64  
 9   Render_Rate   48293 non-null  float64
 10  1st_Clicks    48293 non-null  int64  
 11  2nd_Clicks    48293 non-null  int64  
dtypes: float64(3), int64(5), object(4)
memory usage: 4.4+ MB


In [4]:
gg_df.isnull().sum()

Date               0
App               23
Ad_Group        3993
State              0
Revenue            0
Ad_Requests        0
Ad_Responses       0
Fill_Rate          0
Impressions        0
Render_Rate        0
1st_Clicks         0
2nd_Clicks         0
dtype: int64

1. Thus, the 'App' and the 'Ad_group' has 23 and 3993 null values respetively. 

In [5]:
# group the apps with decreasing order of revenue
app = gg_df.sort_values(['Revenue'],ascending=False).groupby('App').head()
app.head(10)

Date                             App  \
2099  12-09-2020            Bottle Shooting Game   
3650  10-09-2020  Traffic Bangalore: Check Fines   
2084  09-09-2020            Bottle Shooting Game   
2001  08-09-2020            Bottle Shooting Game   
2122  10-09-2020            Bottle Shooting Game   
1953  11-09-2020            Bottle Shooting Game   
3801  11-09-2020  Traffic Bangalore: Check Fines   
4627  09-09-2020  Traffic Bangalore: Check Fines   
4447  08-09-2020  Traffic Bangalore: Check Fines   
4467  07-09-2020  Traffic Bangalore: Check Fines   

                                Ad_Group       State  Revenue  Ad_Requests  \
2099       Bottle Shooting Game_Admob_ML  Unresolved    42.39       416790   
3650  Traffic bangalore_Admob_New Banner   Karnataka    39.34       154116   
2084       Bottle Shooting Game_Admob_ML  Unresolved    34.57       416841   
2001       Bottle Shooting Game_Admob_ML  Unresolved    33.56       428385   
2122       Bottle Shooting Game_Admob_ML  Unresolved    31.86       407421   
1953       Bottle Shooting Game_Admob_ML  Unresolved    31.06       431433   
3801  Traffic bangalore_Admob_New Banner   Karnataka    30.93       139252   
4627  Traffic bangalore_Admob_New Banner   Karnataka    30.09       114676   
4447  Traffic bangalore_Admob_New Banner   Karnataka    28.43       119511   
4467  Traffic bangalore_Admob_New Banner   Karnataka    26.26       118994   

      Ad_Responses  Fill_Rate  Impressions  Render_Rate  1st_Clicks  \
2099        209887      100.0       127509        60.80        8916   
3650        146204       99.8       112777        77.14        1402   
2084        199714      100.0       108858        54.50        7238   
2001        217367      100.0       118674        54.60        7332   
2122        194738       99.9       103268        53.00        7266   
1953        205253      100.0       109628        53.40        7634   
3801        139445       99.9       105400        75.59        1264   
4627        107720       99.7        82558        76.64        1108   
4447        111958       99.8        87418        78.08        1189   
4467        110196       99.6        80531        73.08        1067   

      2nd_Clicks  
2099        2343  
3650           0  
2084        1943  
2001        1944  
2122        2036  
1953        2039  
3801           0  
4627           0  
4447           0  
4467           0

In [6]:
# group the apps with increasing order of revenue
app = gg_df.sort_values(['Revenue'],ascending=True).groupby('App').head()
app

Date                                      App  \
0      08-09-2020                   (SDK X) Cricket Mazza    
32237  09-09-2020                                Baba Cric   
32238  09-09-2020                                Baba Cric   
32239  09-09-2020                                Baba Cric   
32240  09-09-2020  Free turbo VPN - Secure VPN & VPN Proxy   
...           ...                                      ...   
40256  10-09-2020                         Moto Cross Beach   
18301  09-09-2020                   StadtLandFluss (Unity)   
18302  09-09-2020                   StadtLandFluss (Unity)   
22989  08-09-2020                       Mad Crime City 1.0   
22494  06-09-2020                       Mad Crime City 1.0   

                         Ad_Group       State  Revenue  Ad_Requests  \
0      Cricket Mazza Banner (AdX)  Unresolved      0.0      3932330   
32237               Danish&satish   Karnataka      0.0           71   
32238          DanishSolo_English   Karnataka      0.0           71   
32239              Danishsolo_Kan   Karnataka      0.0           71   
32240                     354.857   Rajasthan      0.0           71   
...                           ...         ...      ...          ...   
40256                         NaN  Unresolved      0.0           12   
18301        StadtLandFluss_Admob  Unresolved      0.0         1100   
18302                    revshare  Unresolved      0.0         1100   
22989                         NaN  Unresolved      0.0          428   
22494                         NaN  Unresolved      0.0          472   

       Ad_Responses  Fill_Rate  Impressions  Render_Rate  1st_Clicks  \
0           3926253       99.9      3849749        98.10       10199   
32237             5      100.0            0         0.00           0   
32238             3      100.0            0         0.00           0   
32239             3      100.0            0         0.00           0   
32240            20       97.2           10        50.00           1   
...             ...        ...          ...          ...         ...   
40256             0        0.0            0         0.00           0   
18301           518       54.1            0         0.00           0   
18302            35       54.1           31        88.57           1   
22989             0        0.0            0         0.00           0   
22494             0        0.0            0         0.00           0   

       2nd_Clicks  
0               0  
32237           0  
32238           0  
32239           0  
32240           1  
...           ...  
40256           0  
18301           0  
18302           1  
22989           0  
22494           0  

[1071 rows x 12 columns]

In [7]:
gg_df.App.nunique()

242

Total 242 Unique App

In [8]:
gg_df.Ad_Group.nunique()

188

Total 188 Unique Ad Group

Apps with their total revenue in decreasing order.

In [9]:
# Top 20 Apps with highest sum of Revenue.

gg_df.groupby('App').sum({'Revenue': 'max'})['Revenue'].nlargest(20)

App
Callbreak Multiplayer                                 869.40
Bottle Shooting Game                                  753.89
World Cricket Championship 2                          643.44
Traffic Bangalore: Check Fines                        280.33
World Cricket Championship  Lt                        145.27
Indian Rummy Offline                                  104.28
Tarot Cards Reading and Numerology App                 98.86
Mindi Offline                                          87.34
Indian Train Simulator 2018                            68.51
Sticker Maker                                          61.95
Talk Now – Audio Chat to English speaking practice     61.49
Nursing Exam                                           55.17
(SDK X) Cricket Mazza                                  34.36
Bangla Panjika (বাংলা পঞ্জিকা) 2020 Calendar-1427      31.84
Odia (Oriya) Calendar                                  27.71
Shiva Winter Biking Tales                              23.23
Battle of chepauk   

Callbreak Multiplayer has the highest revenue with $869.40.

In [10]:
# Top 20 Ad Group with highest sum of revenues. 

gg_df.groupby('Ad_Group').sum({'Revenue': 'max'})['Revenue'].nlargest(20)

Ad_Group
Bottle Shooting Game_Admob_ML                    350.40
MY11Circle_IN_SG_1500                            271.80
Callbreak Multiplayer2_Inmobi_No Floor           264.99
IPL banner                                       260.04
Dream11_NonCore                                  236.07
Traffic bangalore_Admob_New Banner               211.05
WCC2_Inmobi_H                                    164.40
Dream11                                          149.45
Rooter_Cricket                                   122.69
Gujrati_GL                                       105.20
4700_Tarot Card_AdMob_Banner                      98.86
WCC2_Admob_GG_0.50                                95.18
Bottle Shooting Game_Admob_No Floor               75.73
non stop fun                                      72.83
Traffic bangalore_Admob_New Banner (backfill)     61.41
RC_MH_English                                     60.84
Nursing_Exam_Admob_Banner_4696                    55.17
sticker Maker (a/b) experiment         

Bottle Shooting Game_Admob_ML has the highest total revenue with $350.40

# Data Manupualtion For Question 1.

#### Let's check the pattern of all the features when the data is sorted by 'Revenue' and grouped by 'App'.

In [11]:
# Apps with with sum of all the numerical features in ascending order.

gg_df.groupby('App').sum('Revenue').sort_values(by='Revenue', ascending=True)

Revenue  Ad_Requests  Ad_Responses  Fill_Rate  \
App                                                                             
Ludo Queen Game                    0.00        16594             0        0.0   
Monster Truck Police Racing        0.00            8             0        0.0   
Monster Truck Racing               0.00           13             0        0.0   
Monster Truck Winter Racing        0.00           26             0        0.0   
Monster Trucks Action Race         0.00            9             0        0.0   
...                                 ...          ...           ...        ...   
World Cricket Championship  Lt   145.27     17363413       1475191   184700.3   
Traffic Bangalore: Check Fines   280.33      1791119       1728060    35502.0   
World Cricket Championship 2     643.44    308410754      15099068   323740.6   
Bottle Shooting Game             753.89    209701332      17484947   231984.5   
Callbreak Multiplayer            869.40   2831311892     131423691   377549.4   

                                Impressions  Render_Rate  1st_Clicks  \
App                                                                    
Ludo Queen Game                           0         0.00           0   
Monster Truck Police Racing               0         0.00           0   
Monster Truck Racing                      0         0.00           0   
Monster Truck Winter Racing               0         0.00           0   
Monster Trucks Action Race                0         0.00           0   
...                                     ...          ...         ...   
World Cricket Championship  Lt      1001420    126878.07      170866   
Traffic Bangalore: Check Fines      1011597     16266.76       11955   
World Cricket Championship 2       12068458    269809.09      763675   
Bottle Shooting Game               13112978    169870.16     1121532   
Callbreak Multiplayer             117137950    314068.94     1092797   

                                2nd_Clicks  
App                                         
Ludo Queen Game                          0  
Monster Truck Police Racing              0  
Monster Truck Racing                     0  
Monster Truck Winter Racing              0  
Monster Trucks Action Race               0  
...                                    ...  
World Cricket Championship  Lt       29033  
Traffic Bangalore: Check Fines           2  
World Cricket Championship 2        149724  
Bottle Shooting Game                330501  
Callbreak Multiplayer               418077  

[242 rows x 8 columns]

From the Above table it is clear that, there are many Apps with zero revenue and all of them share a common pattern. They have few to few thousands Ad Request, but there was no Ad Responses and hence no Fill Rate, Impressions, Render rate, 1St Clicks and 2nd Clicks respectively.

In [12]:
# Apps with with sum of all the numerical features in descending order.

gg_df.groupby('App').sum('Revenue').sort_values(by='Revenue', ascending=False)

Revenue  Ad_Requests  \
App                                                                     
Callbreak Multiplayer                             869.40   2831311892   
Bottle Shooting Game                              753.89    209701332   
World Cricket Championship 2                      643.44    308410754   
Traffic Bangalore: Check Fines                    280.33      1791119   
World Cricket Championship  Lt                    145.27     17363413   
...                                                  ...          ...   
Halloween Cars: Monster Race                        0.00            2   
Halloween Town Racing                               0.00            1   
Hey Flower                                          0.00           61   
Hindi English Translator                            0.00         1560   
Zoo Craft : Blocky World Construction & Builder     0.00         4265   

                                                 Ad_Responses  Fill_Rate  \
App                                                                        
Callbreak Multiplayer                               131423691   377549.4   
Bottle Shooting Game                                 17484947   231984.5   
World Cricket Championship 2                         15099068   323740.6   
Traffic Bangalore: Check Fines                        1728060    35502.0   
World Cricket Championship  Lt                        1475191   184700.3   
...                                                       ...        ...   
Halloween Cars: Monster Race                                0        0.0   
Halloween Town Racing                                       0        0.0   
Hey Flower                                                 56     1700.0   
Hindi English Translator                                  900    18700.0   
Zoo Craft : Blocky World Construction & Builder          2907     2799.4   

                                                 Impressions  Render_Rate  \
App                                                                         
Callbreak Multiplayer                              117137950    314068.94   
Bottle Shooting Game                                13112978    169870.16   
World Cricket Championship 2                        12068458    269809.09   
Traffic Bangalore: Check Fines                       1011597     16266.76   
World Cricket Championship  Lt                       1001420    126878.07   
...                                                      ...          ...   
Halloween Cars: Monster Race                               0         0.00   
Halloween Town Racing                                      0         0.00   
Hey Flower                                                42       885.71   
Hindi English Translator                                 125      2054.75   
Zoo Craft : Blocky World Construction & Builder         2466      2055.62   

                                                 1st_Clicks  2nd_Clicks  
App                                                                      
Callbreak Multiplayer                               1092797      418077  
Bottle Shooting Game                                1121532      330501  
World Cricket Championship 2                         763675      149724  
Traffic Bangalore: Check Fines                        11955           2  
World Cricket Championship  Lt                       170866       29033  
...                                                     ...         ...  
Halloween Cars: Monster Race                              0           0  
Halloween Town Racing                                     0           0  
Hey Flower                                                0           0  
Hindi English Translator                                  1           1  
Zoo Craft : Blocky World Construction & Builder         616         567  

[242 rows x 8 columns]

From the above table it is clear that the Apps with top Revenue have very large number of Ad requests with large number of Ad Responses. When Ad Response is large other features such as Fill Rate, Impressions, Render Rate, 1st Click and 2nd Click are also in large numbers.

## Question 1st
### How would you increase the overall revenue for an app?

## Answer:
### Findings from the above tables - 
1. Top Apps with highest sum of Revenue has high Ad Requests which corresponds to higher values of other features as well.
2. Top apps with lowest sum of Revenue has either very less or zero Ad Requests with zero values for other features.
3. It is clear that the process of revenue generation for an app is a wise sequential method.
4. The sequence goes like this - 
    a) Make high Ad Request which would correspond to a high Ad Response.
    b) High Ad Response would generate high Impression which would correspond to a high Fill Rate and high Render Rate.
    c) When all these parameters are set high there would be high 1st Click and high 2nd Click.
    
5. Thus for increasing the over all revenue of an App, there must be high Ad Request first and then other feature values should also be high.

# Data Manupulation For Question 2.

In [13]:
gg_df.isnull().sum()

Date               0
App               23
Ad_Group        3993
State              0
Revenue            0
Ad_Requests        0
Ad_Responses       0
Fill_Rate          0
Impressions        0
Render_Rate        0
1st_Clicks         0
2nd_Clicks         0
dtype: int64

Thus, there are 23 times when the the App has null values and there are 3993 times when Ad Group has null values. Let's check if the other fetures corresponding to these rows has what values.

In [14]:
gg_df.iloc[gg_df[(gg_df.App.isnull())].index]

Date  App               Ad_Group        State  Revenue  \
48233  07-09-2020  NaN                Dream11  Maharashtra      0.0   
48243  06-09-2020  NaN         Rooter_Cricket  Maharashtra      0.0   
48244  06-09-2020  NaN    RC_MH_Stock_Marathi  Maharashtra      0.0   
48245  06-09-2020  NaN         logo with name  Maharashtra      0.0   
48250  09-09-2020  NaN                Dream11  Maharashtra      0.0   
48251  09-09-2020  NaN         logo with name  Maharashtra      0.0   
48252  09-09-2020  NaN               revshare   Unresolved      0.0   
48253  09-09-2020  NaN         Dhani Backfill   Unresolved      0.0   
48266  08-09-2020  NaN  MY11Circle_IN_SG_1500  Maharashtra      0.0   
48267  08-09-2020  NaN             IPL banner  Maharashtra      0.0   
48268  08-09-2020  NaN               revshare   Unresolved      0.0   
48269  08-09-2020  NaN         Dhani Backfill   Unresolved      0.0   
48281  11-09-2020  NaN             IPL banner  Maharashtra      0.0   
48282  11-09-2020  NaN    RC_MH_Stock_Marathi  Maharashtra      0.0   
48283  11-09-2020  NaN               revshare   Unresolved      0.0   
48284  10-09-2020  NaN             IPL banner  Maharashtra      0.0   
48285  10-09-2020  NaN    RC_MH_Stock_Marathi  Maharashtra      0.0   
48286  10-09-2020  NaN          RC_MH_English  Maharashtra      0.0   
48287  10-09-2020  NaN         logo with name  Maharashtra      0.0   
48288  10-09-2020  NaN               revshare   Unresolved      0.0   
48290  12-09-2020  NaN             IPL banner  Maharashtra      0.0   
48291  12-09-2020  NaN    RC_MH_Stock_Marathi  Maharashtra      0.0   
48292  12-09-2020  NaN               revshare   Unresolved      0.0   

       Ad_Requests  Ad_Responses  Fill_Rate  Impressions  Render_Rate  \
48233            0             0        0.0            0          0.0   
48243            0             0        0.0            0          0.0   
48244            0             0        0.0            0          0.0   
48245            0             0        0.0            0          0.0   
48250            0             0        0.0            0          0.0   
48251            0             0        0.0            0          0.0   
48252            0             0        0.0            0          0.0   
48253            0             0        0.0            0          0.0   
48266            0             0        0.0            0          0.0   
48267            0             0        0.0            0          0.0   
48268            0             0        0.0            0          0.0   
48269            0             0        0.0            0          0.0   
48281            0             0        0.0            0          0.0   
48282            0             0        0.0            0          0.0   
48283            0             0        0.0            0          0.0   
48284            0             0        0.0            0          0.0   
48285            0             0        0.0            0          0.0   
48286            0             0        0.0            0          0.0   
48287            0             0        0.0            0          0.0   
48288            0             0        0.0            0          0.0   
48290            0             0        0.0            0          0.0   
48291            0             0        0.0            0          0.0   
48292            0             0        0.0            0          0.0   

       1st_Clicks  2nd_Clicks  
48233           0           0  
48243           0           0  
48244           0           0  
48245           0           0  
48250           0           0  
48251           0           0  
48252           0           0  
48253           0           0  
48266           0           0  
48267           0           0  
48268           0           0  
48269           0           0  
48281           0           0  
48282           0           0  
48283           0           0  
48284           0           0  
48285      

In [15]:
gg_df.iloc[gg_df[(gg_df.Ad_Group.isnull())].index]

Date                                              App Ad_Group  \
10131  12-09-2020                            Vehicle Owner Details      NaN   
11339  11-09-2020                           Chhota Bheem Race Game      NaN   
11511  09-09-2020                           Chhota Bheem Race Game      NaN   
11547  10-09-2020                           Chhota Bheem Race Game      NaN   
11571  08-09-2020                           Chhota Bheem Race Game      NaN   
...           ...                                              ...      ...   
48228  10-09-2020                                   Barcod Scanner      NaN   
48229  10-09-2020                            Vehicle Owner Details      NaN   
48230  10-09-2020                                       VR Thrills      NaN   
48231  10-09-2020  Zoo Craft : Blocky World Construction & Builder      NaN   
48232  10-09-2020                       Rohit Cricket Championship      NaN   

             State  Revenue  Ad_Requests  Ad_Responses  Fill_Rate  \
10131  Maharashtra      0.0        11174             0        0.0   
11339   Unresolved      0.0         7527             0        0.0   
11511   Unresolved      0.0         7269             0        0.0   
11547   Unresolved      0.0         7193             0        0.0   
11571   Unresolved      0.0         7173             0        0.0   
...            ...      ...          ...           ...        ...   
48228    Telangana      0.0            0             0        0.0   
48229          Goa      0.0            0             0        0.0   
48230  West Bengal      0.0            0             0        0.0   
48231  Maharashtra      0.0            0             0        0.0   
48232    Rajasthan      0.0            0             0        0.0   

       Impressions  Render_Rate  1st_Clicks  2nd_Clicks  
10131            0          0.0           0           0  
11339            0          0.0           0           0  
11511            0          0.0           0           0  
11547            0          0.0           0           0  
11571            0          0.0           0           0  
...            ...          ...         ...         ...  
48228            0          0.0           0           0  
48229            0          0.0           0           0  
48230            0          0.0           0           0  
48231            0          0.0           0           0  
48232            0          0.0           0           0  

[3993 rows x 12 columns]

## Question
### 2. Are there any discrepancies in the data set shared with you? What are the possible reasons for such mismatch?
## Answer
### Findings from Above Tables-
Two types of discripancies were found:
1. There were 23 times when App value was null but still the Ad Groups were available. It seems that for those times the Apps could not make it up to the bidding or did not participated in the bidding.
2. There were 3993 times when App was availabe and in some cases made Ad Request as well but no Ad Group was available (Null Ad Group). It seems that for these times the App had set very low bidding which could not satisfy the Ad Groups. In the cases where the App did not made any Ad Request and therefore no Ad Group were served.



# Data Manupualtion For Question 3.

In [16]:
App_tier = gg_df.groupby('App').sum('Revenue').sort_values(by='Revenue', ascending=False)
App_tier

Revenue  Ad_Requests  \
App                                                                     
Callbreak Multiplayer                             869.40   2831311892   
Bottle Shooting Game                              753.89    209701332   
World Cricket Championship 2                      643.44    308410754   
Traffic Bangalore: Check Fines                    280.33      1791119   
World Cricket Championship  Lt                    145.27     17363413   
...                                                  ...          ...   
Halloween Cars: Monster Race                        0.00            2   
Halloween Town Racing                               0.00            1   
Hey Flower                                          0.00           61   
Hindi English Translator                            0.00         1560   
Zoo Craft : Blocky World Construction & Builder     0.00         4265   

                                                 Ad_Responses  Fill_Rate  \
App                                                                        
Callbreak Multiplayer                               131423691   377549.4   
Bottle Shooting Game                                 17484947   231984.5   
World Cricket Championship 2                         15099068   323740.6   
Traffic Bangalore: Check Fines                        1728060    35502.0   
World Cricket Championship  Lt                        1475191   184700.3   
...                                                       ...        ...   
Halloween Cars: Monster Race                                0        0.0   
Halloween Town Racing                                       0        0.0   
Hey Flower                                                 56     1700.0   
Hindi English Translator                                  900    18700.0   
Zoo Craft : Blocky World Construction & Builder          2907     2799.4   

                                                 Impressions  Render_Rate  \
App                                                                         
Callbreak Multiplayer                              117137950    314068.94   
Bottle Shooting Game                                13112978    169870.16   
World Cricket Championship 2                        12068458    269809.09   
Traffic Bangalore: Check Fines                       1011597     16266.76   
World Cricket Championship  Lt                       1001420    126878.07   
...                                                      ...          ...   
Halloween Cars: Monster Race                               0         0.00   
Halloween Town Racing                                      0         0.00   
Hey Flower                                                42       885.71   
Hindi English Translator                                 125      2054.75   
Zoo Craft : Blocky World Construction & Builder         2466      2055.62   

                                                 1st_Clicks  2nd_Clicks  
App                                                                      
Callbreak Multiplayer                               1092797      418077  
Bottle Shooting Game                                1121532      330501  
World Cricket Championship 2                         763675      149724  
Traffic Bangalore: Check Fines                        11955           2  
World Cricket Championship  Lt                       170866       29033  
...                                                     ...         ...  
Halloween Cars: Monster Race                              0           0  
Halloween Town Racing                                     0           0  
Hey Flower                                                0           0  
Hindi English Translator                                  1           1  
Zoo Craft : Blocky World Construction & Builder         616         567  

[242 rows x 8 columns]

In [17]:
pd.DataFrame(App_tier.Revenue>100).head()

Revenue
App                                    
Callbreak Multiplayer              True
Bottle Shooting Game               True
World Cricket Championship 2       True
Traffic Bangalore: Check Fines     True
World Cricket Championship  Lt     True

In [18]:
pd.DataFrame(App_tier.Revenue>100).value_counts()

Revenue
False      236
True         6
dtype: int64

## Question
### 3. If you had a classification mechanism, which apps would you classify as Tier 1 or Important category for GG? Explain your reasoning.
## Answer
### Findings from Above Tables-
Thus from the above tables we can see that there only six Apps which has Revenue greater than $100. These Apps could be classified as 'Tier 1 App'.

# Data Manupualtion For Question 4.

In [19]:
Ad_Group_Rev = gg_df.groupby('Ad_Group').sum('Revenue').sort_values(by='Revenue', ascending=True)
Ad_Group_Rev

Revenue  Ad_Requests  Ad_Responses  \
Ad_Group                                                                     
test                                       0.00            7             4   
Live Talk_Admob_MCM                        0.00      1638798        532430   
Latest Attitude Status 2020_Admob          0.00         2013           331   
Kannada Keyboard_Inmobi                    0.00           16             2   
Jesus Wallpapers                           0.00          206            81   
...                                         ...          ...           ...   
Dream11_NonCore                          236.07    161165324      54420311   
IPL banner                               260.04    158138905      21087660   
Callbreak Multiplayer2_Inmobi_No Floor   264.99    131428733       3534284   
MY11Circle_IN_SG_1500                    271.80    160508752      21562112   
Bottle Shooting Game_Admob_ML            350.40     17485304       3727076   

                                        Fill_Rate  Impressions  Render_Rate  \
Ad_Group                                                                      
test                                        400.0            0         0.00   
Live Talk_Admob_MCM                       21998.6       505478     20946.93   
Latest Attitude Status 2020_Admob          8700.0            0         0.00   
Kannada Keyboard_Inmobi                     200.0            0         0.00   
Jesus Wallpapers                           3725.0            0         0.00   
...                                           ...          ...          ...   
Dream11_NonCore                           84820.9     50103937     72247.75   
IPL banner                               153426.3     18362860    123796.62   
Callbreak Multiplayer2_Inmobi_No Floor    24197.0      2828252     19325.25   
MY11Circle_IN_SG_1500                    155019.8     18646537    125279.41   
Bottle Shooting Game_Admob_ML             24398.4      2090511     13011.51   

                                        1st_Clicks  2nd_Clicks  
Ad_Group                                                        
test                                             0           0  
Live Talk_Admob_MCM                         419120       20347  
Latest Attitude Status 2020_Admob                0           0  
Kannada Keyboard_Inmobi                          0           0  
Jesus Wallpapers                                 0           0  
...                                            ...         ...  
Dream11_NonCore                             488733      120575  
IPL banner                                  469666      122258  
Callbreak Multiplayer2_Inmobi_No Floor       82826       26686  
MY11Circle_IN_SG_1500                       519413      133980  
Bottle Shooting Game_Admob_ML               128963       35054  

[188 rows x 8 columns]

In [20]:
pd.DataFrame(Ad_Group_Rev.Revenue==0).head()

Revenue
Ad_Group                                  
test                                  True
Live Talk_Admob_MCM                   True
Latest Attitude Status 2020_Admob     True
Kannada Keyboard_Inmobi               True
Jesus Wallpapers                      True

In [21]:
pd.DataFrame(Ad_Group_Rev.Revenue==0).value_counts()

Revenue
False      103
True        85
dtype: int64

## Question
### 4. Are there any ad groups which are causing an opportunity loss for the app? If yes, how would you mitigate it? 
## Answer
### Findings from Above Tables-
1. There are total of 85 Ad group with sum of zero revenue, even in some cases Ad Requests were made by the App.
2. To mitigate this loss, the bidding could be brought down optimally low which would encourage the App to participate in the bidding.